### DISCUSSION TITLE
Is there a reason we can't just import all devices supported by OpenWRT?

### DISCUSSION THREAD

**Discussion by Darin755** [openwisp/openwisp-firmware-upgrader#264] [Ideas]
If I were to create a PR with 1000 new devices would it be accepted? I don't own all of those devices but I can pull them off the table.

**Comment by nemesifier** [comment]
Hi @Darin755, thanks for your question. I am honestly not sure right now.

The list of devices changes constantly, some are added, some are removed. Some definitions and labels change in each OpenWrt version.

If we add thousands of lines of codes but nobody is using those lines of codes.. who is it going to maintain them?

If we want to do something like this, it may be more appropriate to look for a way to automate this so we can be confident the list stays up to date over time and can work with different OpenWrt versions, but this is not a trivial challenge. It would be great if the OpenWrt project offered some sort of structured data we could pull periodically to build our list. Maybe they do and we are not aware of that?

Let me ask you a few questions to understand your point of view. Why are you prompted to do this?
Are you using this module and often need to add more devices, which are not included in the list?

**Reply by Darin755** [reply]
The OpenWRT project offers a CSV hardware dump. I figured it might be possible to create a automated pull request once a month to update the hardware list. 

https://openwrt.org/supported_devices

**Reply by nemesifier** [reply]
That is great, we have to verify if any of those columns corresponds to the values we need to add in "boards" here:
https://github.com/openwisp/openwisp-firmware-upgrader/blob/master/openwisp_firmware_upgrader/hardware.py

**Reply by Darin755** [reply]
That might be tricky but I think it is doable

**Reply by Darin755** [reply]
Would you be ok accepting a big PR with a giant commit or would you like me to break it down into smaller chunks?

**Reply by nemesifier** [reply]
@Darin755 I think a giant commit is fine as long as we have the following:

- Separate the definitions for the openwrt images in a dedicated directory, we could have one file per soc to avoid having one giant scary file, we can have the hardware.py import everything, or something along these lines.
- A script to reproduce the result and update the file, we will add a mention of this script in the docs for maintenance purposes.
- How do we deal with the same hardware defined both as old ar71xx and new ath79 target? I am not sure right now.

**Reply by Darin755** [reply]
I was looking to maybe just have the CSV from OpenWRT in its own directory. It might be wasteful to not strip down the CSV with only needed information but I figured having a database of all supported hardware might be useful in the future. It would also simplify updating it as a updater script could just use curl to pull from the website. I can imagine a future where there might be an about button that shows basic information about the specs of a device.

Let me know what you think as I am still pondering the design off this. I realize having a giant CSV might not be appealing but I will make sure to document the code with comments and I might even add a readme with formatting information. My fear is that some time down the road the OpenWRT CSV will change format.

**Reply by nemesifier** [reply]
@Darin755, we'll need to store information in a python data structure, having it as CSV will be redundant. 
I don't think it's useful to include the CSV in the repository. Having the script which parses the CSV and converts it to Python datastructure is.
Yes the format may change, as many things with OpenWrt. However, we can try to get in touch with the people who have put effort to produce that CSV and communicate to them what we're doing and let them know that it would be great to ensure some sort of stability.
Do you know who's responsible for that? I'll ask to a few devs I know on my end.

We will also have to see how the system reacts after having such a huge data structure loaded in, for example, I am wondering if RAM usage would increase significantly? Hopefully not. But if we find constraints in Django model choices, we may need to rethink this and probably convert this information as its own DB table, but this is overkill for now: lots of work for little added benefit IMHO. 
What OpenWISP needs more right now is to remove friction on its first steps: making installation and setup of common scenarios easier, improve communication of what it does and its documentation, that's what we've been focused on lately.

Have you been able to use the settings provided in this repo to do what you needed? Or was there a problem with that? 

I think the first step would be to try loading the data set in a dirty way, just to see what happens, if the system can handle it.
I am not sure Django model choices were designed to be huge.

**Reply by nemesifier** [reply]
Copying the messages from an OpenWrt dev verbatim:

> I don't even know that CSV
> There is the Table of Hardware (ToH) on the Wiki
> maybe it's a dump of that?
> and the wiki, well, it's a wiki, the users are responsible for that ;)
> and then there is profiles.json for each subtarget
> imho that's the most authoritative source because it really all the boards we generate an image for, not more and not less
> https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/profiles.json
> https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/profiles.json
> (and so on...)

Now, I am taking a look at the mt7621 JSON and I have already found something.

Look at this model:
https://github.com/openwrt/openwrt/blob/d604b2699b6015f99ff194b34f532958bde7d442/target/linux/ramips/dts/mt7621_zbtlink_zbt-wg3526-16m.dts#L4

The model line says: `model = "Zbtlink ZBT-WG3526 (16M)";`.

This is what we have in hardware.py.

However, the json mentions the following:

```
"titles": [
        {
          "model": "ZBT-WG3526",
          "variant": "16M",
          "vendor": "Zbtlink"
        }
```

Does it mean we have to construct it as `{vendor} {model} ({variant})`? It may be, but we have to triple check.

And what happens if they change it in the source code? We need to support multiple versions of OpenWrt, not just one version.
That's why in `hardware.py` the `'boards` dict key expects an iterable (list or tuple), because different versions of OpenWrt may use slightly different board names.

Why is it important to have the correct board names in `hardware.py`? Because when devices register, we want to find out if the device is equipped with one of the images which are stored in the fw-upgrader module and in that case we want to create a relation between the device and the image, so that when a new version of the image is available on the system, users can upgrade easily.

Doing this integration is definitely doable but without exercising some pressure on the OpenWrt development it may be futile. Or we need the script to take into account these possible changes and generate an `hardware.py` which is the result of the changes from multiple versions. 

Another way of doing it would be to instruct the script to look directly in the image definitions `dts` files and extrat the information from there. This would probably be more reliable.

**Reply by Darin755** [reply]
Would it be a fair statement that OpenWISP only will support OpenWRT versions that are not EOL?

**Reply by nemesifier** [reply]
> Would it be a fair statement that OpenWISP only will support OpenWRT versions that are not EOL?

It's not black and white: it depends on a case by case basis.
Moreover, removing a definition from hardware.py on a system that uses that definition is going to crash it and that's definitely something we can't allow without providing a gradual deprecation/upgrade path.

**Reply by Darin755** [reply]
Maybe I could create a json with current supported devices and then have hardware.py load it

**Reply by nemesifier** [reply]
What reason is prompting you to move the data from the python file, which is the native language of this application, to a JSON file?
What advantages do you see? I think most people contributing to this repository are comfortable working with Python, why alienate them?

Or do you mean generate JSON from the data in OpenWrt and then convert that into a Python data structure loaded in hardware.py?

**Reply by Darin755** [reply]
I am trying to understand what is needed by this project. You said that you wanted to be sure that there are no breaking changes and I figured adding a ton of new untested devices could be a problem. Therefore I thought it might be optimal to take the existing devices and put then into a json. The reason I suggest json is that it is easy to generate json and I could create a script to add more devices down the road.

**Reply by okraits** [reply]
The questions which come to my mind:

* What are the criteria for adding/removing device models? Popularity, used by OpenWISP project members, added/removed by PR's, etc?
* Do we really want to have a list with 1000 device models? I can imagine this would get confusing in terms of usability when searching/selecting a device...

Another thing to consider is that OpenWISP might rather be used by organizations and less by individuals. And organizations might have a limited set of device models which they can add with `OPENWISP_CUSTOM_OPENWRT_IMAGES`. So it might not make much sense to include a list of 1000 device models, anyway.

**Reply by Darin755** [reply]
That is what I am trying to figure out. I do like the way the OpenWRT firmware selector works.

https://firmware-selector.openwrt.org/

Maybe we could add all devices but instead of just listing them in a giant drop you could add a search with search suggestions. Doing that would take a little be of doing so putting the currently supported devices in a json for now might be a good way to future proof. 

It really depends on what you guys want. I am happy to help and I would love to see all devices supported for convenience reasons.

**Reply by nemesifier** [reply]
> The questions which come to my mind:
> 
> * What are the criteria for adding/removing device models? Popularity, used by OpenWISP project members, added/removed by PR's, etc?

The criteria for devices listed in `hardware.py` should be devices which are supported by OpenWrt or were supported before (and OpenWISP also supported those here, so removing them would cause issues to existing installations using those devices).

> * Do we really want to have a list with 1000 device models? I can imagine this would get confusing in terms of usability when searching/selecting a device...

In the latest version we're already using the select autocomplete widget to avoid having problems with the length of the devices.
But we'd have to see if having a giant list would cause problems elsewhere. In that case we'd have to rewrite this part.
 
> Another thing to consider is that OpenWISP might rather be used by organizations and less by individuals. And organizations might have a limited set of device models which they can add with `OPENWISP_CUSTOM_OPENWRT_IMAGES`. So it might not make much sense to include a list of 1000 device models, anyway.

This is the reason for which it hasn't been a pressing issue. Nonetheless, having away to automate the compilation and maintenance of the list would surely be helpful, provided we keep in mind backward compatibility.

**Reply by Darin755** [reply]
So do you want new devices in hardware.py exclusively? I am trying to figure out how I should implement this. I first choice would be a json but that can change based on what kind of changes you would accept.

**Reply by nemesifier** [reply]
It's better to use python files, we already have a lot of logic and QA checks to enforce readability and correctnes.
Instead of one giant file, we could have one file for each target and then import/load everything in one file which is then loaded from hardware.py.

In the future, hardware.py will load definitions for multiple systems, not only openwrt.

**Reply by Darin755** [reply]
Thanks that makes sense. I'll look into it
